[needs 5758] Qt: replace EmulationStateChanged connections with Core callbacks

ligfx opened the pull request:
Removes the soup of `connect(parent, &Parent::EmulationStarted, child, &Child::EmulationStarted)` that was needed to pass signals through multiple levels of widgets. Blegh.

Other changes:
- Adds a `QueueOnObject` function in QtUtils, to run code on the correct thread for a QWidget.
- Adds a `ConnectToSubscribable` function in QtUtils, to hook up callbacks that will be automatically unsubscribed when a widget is destroyed (turns out you can't connect to an already-destroyed widget without crashing).

spycrab commented:
👍 this concept. Required for fixing a current bug.

spycrab commented:
Forgot to ask: Is there any Qt slot binding for ``Core::AddOnStateChangedCallback`` yet? In ``Settings.cpp`` perhaps?

ligfx commented:
Woops, this currently causes a segfault when quitting while a game is running. I'm looking into it.

@spycrab I didn't put in a Qt signal. In the code I'm doing:

```c++
Core::AddOnStateChangedCallback([this](Core::State state) { QueueOnObject(this, [this, state]{
  // do some stuff
});});
```
which handles the "run logic on the correct thread" and "don't segfault if object was destroyed already" parts of Qt signals (`QueueOnObject` is using signals internally).

ligfx commented:
I pulled the modifications to `Core` out into a separate PR and rebased this one on top of it. The new implementation supports unsubscribing callbacks, which is helpful for preventing the segfaults (turns out you can't call `QObject::connect` on an already-destroyed object).

@spycrab I added a helper, so connecting to the state change notifications now looks like:

```c++
#include "Core/Core.h"
#include "DolphinQt2/QtUtils/ConnectToSubscribable.h"

ConnectToSubscribable(Core::g_on_state_changed, this, [=](Core::State state) {
  // do some stuff, guaranteed to run on correct thread and with a valid `this`
});
```

spycrab commented:
@ligfx Looks really neat! Can't wait to putting that to good use.
